src/components/NewDesign/Swap/Swap.tsx

```diff
@@ -0,0 +1,379 @@
+    v: BN
+  }
+  fee: {
+    val: BN,
+    scale: number
+  }
+  exchangeRate: {
+    val: BN,
+    scale: number
+  }
+}
+
+export interface ISwap {
+  walletStatus: Status
+  tokens: SwapToken[]
+  pools: Pools[]
+  onSwap: (fromToken: PublicKey, toToken: PublicKey, amount: BN) => void
+}
+export const Swap: React.FC<ISwap> = ({
+  walletStatus,
+  tokens,
+  pools,
+  onSwap
+}) => {
+  const classes = useStyles()
+  const [tokenFromIndex, setTokenFromIndex] = React.useState<number | null>(
+    tokens.length ? 0 : null
+  )
+  const [tokenToIndex, setTokenToIndex] = React.useState<number | null>(null)
+  const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null)
+  const [amountFrom, setAmountFrom] = React.useState<string>('')
+  const [amountTo, setAmountTo] = React.useState<string>('')
+  const [swap, setSwap] = React.useState<boolean | null>(null)
+  const [priceProportion, setPriceProportion] = React.useState<BN>(new BN(1))
+  const [tokenY, setTokenY] = React.useState<SwapToken[] | null>(null)
+  const [poolIndex, setPoolIndex] = React.useState<number | null>(null)
+  const [slippTolerance, setSlippTolerance] = React.useState<string>('')
+  const [settings, setSettings] = React.useState<boolean>(false)
+  const [details, setDetails] = React.useState<boolean>(false)
+  // const firstUpdate = useRef(true)
+  const calculateSwapOutAmount = (assetIn: SwapToken, assetFor: SwapToken, amount: string) => {
+    // TODO: solution if 0 => change to 1
+    if (priceProportion.eqn(0)) {
+      setPriceProportion(new BN(1))
+    }
+
+    const amountOut: BN =
+      swap
+        ? printBNtoBN(amount, assetIn.decimal).mul(priceProportion)
+        : printBNtoBN(amount, assetIn.decimal).div(priceProportion)
+
+    return printBN(amountOut, assetFor.decimal)
+  }
+
+  useEffect(() => {
+    updateEstimatedAmount()
+
+    if ((tokenFromIndex !== null && tokenToIndex === null)) {
+      setAmountFrom('0.000000')
+    }
+    if (tokenFromIndex !== null) {
+      const tokensY = tokens.filter((token) => {
+        if (swap) {
+          return getSwapPoolIndex(token.assetAddress, tokens[tokenToIndex ?? 0].assetAddress) !== -1
+        } else {
+          return getSwapPoolIndex(token.assetAddress, tokens[tokenFromIndex].assetAddress) !== -1
+        }
+      })
+      setTokenY(tokensY)
+    }
+  }, [tokenToIndex, tokenFromIndex])
+
+  useEffect(() => {
+    if (tokenFromIndex !== null && tokenToIndex !== null) {
+      const pairIndex = pools.findIndex((pool) => {
+        return (tokens[tokenFromIndex].assetAddress.toString() === pool.tokenX.toString() &&
+        tokens[tokenToIndex].assetAddress.toString() === pool.tokenY.toString()) ||
+       (tokens[tokenToIndex].assetAddress.toString() === pool.tokenX.toString() &&
+        tokens[tokenFromIndex].assetAddress.toString() === pool.tokenY.toString())
+      })
+      if (pairIndex !== -1) {
+        setPriceProportion(pools[pairIndex].sqrtPrice.v.div(new BN(10 ** PRICE_DECIMAL))
```

Comment:
what if price proportion in pools array will be updated? this might not update

Reply:
done